# 1 Author

**Student Name**:  Yashika

**Student ID**:  220802299



# 2 Problem formulation

Here, we have divided the spots in the daatset in 3 categories based on some parameters such as noise, number of people present at the place and the type of place(whether speaking loudly is permissible at the place).
The categories are:

**Cat_A**:library, forecourt, reception, national, bridge, gardens, royal, pond, room13, canal, room12

**Cat_B**:square, ground, street, greatcourt, upper, ritblat, albert, dinosaur

**Cat_C**:piazza, trafalgar, leicester, curve, charing, market, food, hintze, skate, book, waterloo, cromwell, marine

We can see that the spots in Cat_A are least noisy, so background noise should be minimum at these places.

The spots in Cat_B are areas with a medium level of noise.

The spots in Cat_C are suspected to have been most crowded and with more background chattering and noises.


Using the MLEnd London Sounds dataset, we want to train a machine learning model that takes as an input an audio segment and predicts whether the audio segment has been recorded at one of these places(Cat_a/Cat_B/Cat_C). We will then be evaluating the accuracy of the model on the training and test data.

# 3 Machine Learning pipeline


Following steps will be carried out:

1. First of all, we will extract a sample(of length 1000) of audio files into a location that can be accessed by the program.
2. Next, we will load the csv file containing the information of the Sounds dataset(attributes and features) into a pandas dataframe.
3. Now we need to identify what features are relevant to classify an audio file as being recorded at one of the 3 places from Cat_A, Cat_B, Cat_C. Feature selection is a very important step as the prediction depends on this. If the features are not selected carefully, there is a chance for the model to underfit/overfit. The features selected for this problem statement are:
  1. Energy
  2. Zero Crossing Rate
  3. Signal to noise raio
  4. Flatness
4. Next, we will define a function to calculate the values of these features.
5. Once the functions are defined, we will iterate over the set of audio files to execute the function to get the values of these features for each audio.
After this step, we will have 2 arrays X and Y where X will contain the features of the audio files in the form of some values and Y will contain the actual label of the audio file (Cat_A, Cat_B, Cat_C). We will also encode the labels using LabelEncoder. It will encode target labels with value between 0 and n_classes-1.
6. Once we have the datasets, we need to select the model that we need to train. Different types of models can be accurate for different types of datasets. We will analyse the accuracy of the training and test data using different models and will choose the final model according to that.
7. Now, we will split our dataset into 2 parts i.e. train and test. We will be using 70% of the datset for training and the remaining 30% for testing.
8. Now when we have our datasets ready, we will train the selected model by passing into it the training datset.
9. Once this has been done, we will predict the training and validation labels using the trained model and calculate the accuracy for each.
10. We can check the change in accuracy by running a number of iterations for steps 7-9.
11. Next, we can normalise our dataset and try steps 7-10 again to see if the accuracy improves.
12. Finally, we will analyse the accuracy of different models and select the final model based on the results and some facts.

# 4 Transformation stage


Since our data was in the form of audio files, we could not use it directly to make any predictions. Hence we have to perform transformation on our datset to get relevant features out of it. The input to this step is the set of audio files and the output is a 2D array that contains 4 features for each audio. Each row in X will correspond to an audio file and each column will be a feature as described above. The resulting dataset after transformation will be of the form:

X(features): (1000,4)

Y(labels): (1000,)

The reason for selecting each of the features is:

1. Energy - It provides a way to idetify that a sound is a tone or more like noise. This can be helpful in identifying the label as there is likely to be more noise in a place which is more populated than the others.

2. Zero Crossing rate - The zero-crossing rate is the rate at which a signal changes from positive to zero to negative or from negative to zero to positive. It can be a helpful measure in classifying the samples in our problem statement because the chattering or chirping noises in the background will constitute to more crossingss of signal through the horizontal axis. Thus it can classify the places as being more or less noisy.

3. Signal to noise ratio - It can be a useful measure to classify an audio as being recorded at one of the 3 places(Cat_A, Cat_B, Cat_C) as it tells about the performance of an audio signal in terms of noise and signal quality. It compares the strength of the audio signal to the level of noise in the audio. Again, the more populated places might have more background noises, which can be used as a classifier.

4. Flatness - It provides a way to identify that a sound is a tone or more like noise. This can be helpful in identifying the label as there might be more noise in an audio recorded near a skate stadium than in a library.

# 5 Modelling

Here, we will train multiple models i.e. KNN, Random Forest Classifier and Logistic Regression model. We will not be using SVM here because SVM model is best used to classify binary values. We will analyse the accuracy of our training and test data with each of these models and will choose the appropriate model at the end based on certain results and facts.

# 6 Methodology

We will divide the X(features) and Y(actual labels) array into 2 parts i.e. train and test. 70% of the data will be used to train the model and the rest 30% will be used to test the model. We will then measure the training and validation accuracy by running a number of iterations. We need to be careful that model should not overfit/underfit in any scenario. We will assess the accuracy of the model mainly by 2 factors:
  1. Accuracy on train and test data(both before and after normalisation)
  2. Whether the model is underfitting/overfitting



# 7 Dataset

We have the following data to train and test our model as per the problem statement:
  1. Sample of 1000 audio files
  2. A CSV file that contains the following information about each audio:
    * **file_id** - name of the audio file, which will be same as the audio name that we have already extracted.
    * **area** - The area where the audio file has been recorded. It can contain the following 6 values:
      * british
      * kensington
      * campus
      * westend
      * Euston
      * southbank
    * **spot** - The particular spot in that area where the audio has been recorded. Each of the 6 areas can have 6 different values for the spot. These are described as below:

          british - street, forecourt, greatcourt, room12, square, room13

          campus - canal, curve, ground, library, reception, square

          Euston - forecourt, gardens, library, piazza, ritblat, upper

          kensington -  albert, cromwell, dinosaur, hintze, marine, pond

          southbank - book, bridge, food, royal, skate, waterloo

          westend - charing, leicester, market, national, piazza, trafalgar
    * **in_out** - This column can contain 2 different values:
      * indoor - If the audio file has been recorded indoors
      * outdoor - If the audio file has been recorded outdoors
    * **Participant** - The task of recording audios was carried out by various participants. This column contains the participant id of the person who has recorded the audio.


We cannot use the audio files directly to train our model. Thus we need to extract some features out of it that will be further passed as input to the model. Functions have been defined to extract the features and the code for feature extraction is shown below. 

Also, we need to load our csv file in a pandas dataframe, so that we can perform operations on it.

The output of the Transformation step will be 2 2-Dimensional arrays:

X(Array of attributes) - Shape of this will be (1000,4) where each row will represent an audio file and each column will correspond to a feature.

Y(array of labels) - Shape of this will be (1000,) and this array will contain the actual labels of the corresponding entries in X. We will return this value in encoded form and it will take values 0,1,2 - one for each category(Cat_A, Cat_B, Cat_C)





In [4]:
from google.colab import drive

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import os, sys, re, pickle, glob
import urllib.request
import zipfile

import IPython.display as ipd
from tqdm import tqdm
import librosa
import scipy.io.wavfile as wavfile
import os.path
import scipy.stats as stats
import scipy

drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
sample_path = '/content/drive/MyDrive/Data/MLEndLS/sample/*.wav'
files = glob.glob(sample_path) # Storing all the files matching the pattern oin a variable
print("Total number of audio files is: ", len(files))

Total number of audio files is:  1000


In [6]:
MLENDLS_df = pd.read_csv('./MLEndLS.csv').set_index('file_id') # Loading csv to dataframe
MLENDLS_df

area       spot   in_out Participant
file_id                                             
0001.wav     british     street  outdoor        S151
0002.wav  kensington   dinosaur   indoor        S127
0003.wav      campus     square  outdoor         S18
0004.wav  kensington     hintze   indoor        S179
0005.wav      campus     square  outdoor        S176
...              ...        ...      ...         ...
2496.wav     westend  trafalgar  outdoor        S151
2497.wav      campus     square  outdoor          S6
2498.wav     westend   national   indoor         S96
2499.wav     british     room12   indoor         S73
2500.wav     british     room12   indoor         S58

[2500 rows x 4 columns]

In [13]:
cat_dict = {"Cat_A":["library","forecourt","reception","national","bridge","royal","pond","gardens","room13","canal","room12"], "Cat_B":["square","ground", "street","greatcourt","upper","ritblat","albert","dinosaur"], "Cat_C":["piazza","trafalgar","leicester","curve","charing","market","food","hintze","skate","book","waterloo","cromwell","marine"]}
MLENDLS_df2 = MLENDLS_df.assign(Category="")
for i, row in MLENDLS_df2.iterrows():
  row["Category"] = [k for k, v in cat_dict.items() if row["spot"] in v][0]
MLENDLS_df2

area       spot   in_out Participant Category
file_id                                                      
0001.wav     british     street  outdoor        S151    Cat_B
0002.wav  kensington   dinosaur   indoor        S127    Cat_B
0003.wav      campus     square  outdoor         S18    Cat_B
0004.wav  kensington     hintze   indoor        S179    Cat_C
0005.wav      campus     square  outdoor        S176    Cat_B
...              ...        ...      ...         ...      ...
2496.wav     westend  trafalgar  outdoor        S151    Cat_C
2497.wav      campus     square  outdoor          S6    Cat_B
2498.wav     westend   national   indoor         S96    Cat_A
2499.wav     british     room12   indoor         S73    Cat_A
2500.wav     british     room12   indoor         S58    Cat_A

[2500 rows x 5 columns]

In [7]:
def signaltonoise(a, axis=0, ddof=0):
    a = np.asanyarray(a)
    m = a.mean(axis)
    sd = a.std(axis=axis, ddof=ddof)
    return np.where(sd == 0, 0, m/sd)
def snr(file):
  data = wavfile.read(file)[1]
  singleChannel = data
  try:
    singleChannel = np.sum(data, axis=1)
  except:
    pass
    
  norm = singleChannel / (max(np.amax(singleChannel), -1 * np.amin(singleChannel)))
  return signaltonoise(norm)

In [8]:
def getEnergy(x,fs,winLen=0.02):
  p = winLen*fs
  frame_length = int(2**int(p-1).bit_length())
  hop_length = frame_length//2
  rmse = librosa.feature.rms(x, frame_length=frame_length, hop_length=hop_length, center=True)
  return rmse

In [82]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
def get_features_labels(files,labels_file, scale_audio=False, onlySingleDigit=False):
  X,y =[],[]
  for file in tqdm(files):
    try:
      fileID = file.split('/')[-1]
      file_name = file.split('/')[-1]
      yi = labels_file.loc[fileID]['Category']

      fs = None # if None, fs would be 22050
      x, fs = librosa.load(file,sr=fs)
      energy = np.mean(getEnergy(x,fs,winLen=0.02))
      flatness = np.mean(librosa.feature.spectral_flatness(y=x))
      zcr = np.mean(librosa.feature.zero_crossing_rate(x))
      snratio = snr(file)
      xi = [energy, zcr, snratio, flatness]
      X.append(xi)
      y.append(yi)
    except Exception as e:
      print("Broken file: ", file, e)
  y1 = le.fit_transform(y)

  return np.array(X),np.array(y1)

In [83]:
X,y = get_features_labels(files, labels_file=MLENDLS_df2, scale_audio=True, onlySingleDigit=True)

100%|██████████| 1000/1000 [01:31<00:00, 10.91it/s]


In [84]:
print('The shape of X is', X.shape) 
print('The shape of y is', y.shape)
print('The labels vector is', y)

The shape of X is (1000, 4)
The shape of y is (1000,)
The labels vector is [0 0 0 1 0 2 0 1 1 0 1 0 1 0 1 2 2 2 0 1 1 0 1 0 1 2 1 0 0 2 2 2 0 1 1 2 1
 0 1 1 2 1 0 2 1 0 0 2 2 2 1 1 0 0 2 0 0 0 1 0 2 0 2 1 2 2 0 0 2 1 2 1 1 0
 2 0 0 2 0 1 2 0 0 2 1 0 1 2 0 1 1 0 2 1 2 0 2 2 2 2 2 2 2 2 1 2 2 2 1 0 0
 0 0 2 0 0 2 1 0 2 1 2 1 1 1 2 1 2 0 1 1 1 0 1 2 2 1 2 2 1 0 1 1 2 1 0 0 1
 0 2 1 1 0 1 1 2 2 1 2 2 0 2 1 1 1 1 2 1 2 2 1 2 0 0 0 0 2 2 2 2 2 2 1 1 0
 2 1 2 2 0 2 2 0 2 0 1 2 2 1 1 1 0 1 0 0 2 2 2 0 1 2 2 2 1 0 1 1 0 2 2 0 1
 2 0 1 1 0 0 0 2 0 0 2 1 2 0 2 0 0 2 0 2 2 0 1 0 1 1 1 0 0 2 0 0 1 0 0 1 2
 2 1 0 2 0 2 0 2 2 2 1 2 0 1 2 2 1 1 0 1 1 1 1 2 2 2 0 2 2 0 2 0 1 0 2 1 1
 0 0 1 1 2 1 2 0 1 0 2 2 0 0 2 1 0 0 0 0 0 0 2 1 2 1 2 1 0 2 2 2 2 1 2 1 2
 2 1 2 0 1 0 2 2 0 0 0 2 1 2 2 0 0 0 2 0 2 2 0 0 2 1 2 2 2 1 2 1 0 2 2 2 0
 1 2 0 0 2 2 0 2 0 0 2 2 2 1 2 2 1 0 0 0 0 0 0 2 0 0 0 2 2 0 0 0 2 0 0 1 2
 0 0 1 0 1 2 0 1 1 0 0 0 0 0 2 2 1 1 2 0 0 1 2 0 2 2 0 2 1 0 0 2 1 0 2 0 2
 2 1 1 0 0 2 0 2 2 1 2 2 

In [85]:
unique, count = np.unique(y, return_counts=True)
print(' The count for each label is:\n ', np.asarray((unique, count)).T)

 The count for each label is:
  [[  0 359]
 [  1 272]
 [  2 369]]


# 8 Results

Now, when we have our arrays of attributes and labels ready, we need to carry out the main task of training the mdoel and analysing the training and validation accuracy to select the right model. We will be doing it here.

We will train 3 different models of supervised learning:


  **k nearest neighbour** - This is based on the notion that the features of a datapoint can be predicted based on the features of its neighbours.

  **Logistic Regression Model** - The binary logistic regression model is used in the cases where there are 2 classes. It is mostly used with categorical classes.

  **Random Forest Classifier** - This model is efficient in classifying large datsets. the accuracy of this model is better than a decision tree because it uses a collection of decision trees, and each tree in the ensemble is comprised of a data sample drawn from a training set with replacement.

In [95]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

X_train, X_val, y_train, y_val = train_test_split(X,y,test_size=0.3)
print("The shape of training set(X) is : {}\nThe shape of validation set(X) is : {}\nThe shape of training set(Y) is : {}\nThe shape of validation set(Y) is : {}".format(X_train.shape, X_val.shape, y_train.shape, y_val.shape))

The shape of training set(X) is : (700, 4)
The shape of validation set(X) is : (300, 4)
The shape of training set(Y) is : (700,)
The shape of validation set(Y) is : (300,)


In [92]:
model  = KNeighborsClassifier()
model.fit(X_train,y_train)
yt_p = model.predict(X_train)
yv_p = model.predict(X_val)

train_labels = np.unique(y_train)
val_labels = np.unique(y_val)

print("\n\n--------FOR KNN MODEL-------")
print('Training Accuracy', np.mean(yt_p==y_train))
print('Validation  Accuracy', np.mean(yv_p==y_val))
conf_matrix_train = metrics.confusion_matrix(y_train, yt_p, labels=train_labels)
conf_matrix_val = metrics.confusion_matrix(y_val, yv_p, labels=val_labels)

print('Training confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_train, index=train_labels, columns=train_labels)))
print('Validation confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_val, index=val_labels, columns=val_labels)))

model = LogisticRegression()
model.fit(X_train,y_train)
yt_p = model.predict(X_train)
yv_p = model.predict(X_val)

print("\n\n--------FOR LOGISTIC REGRESSION MODEL-------")
print('Training Accuracy', np.mean(yt_p==y_train))
print('Validation  Accuracy', np.mean(yv_p==y_val))
conf_matrix_train = metrics.confusion_matrix(y_train, yt_p, labels=train_labels)
conf_matrix_val = metrics.confusion_matrix(y_val, yv_p, labels=val_labels)

print('Training confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_train, index=train_labels, columns=train_labels)))
print('Validation confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_val, index=val_labels, columns=val_labels)))

model = RandomForestClassifier()
model.fit(X_train,y_train)
yt_p = model.predict(X_train)
yv_p = model.predict(X_val)

print("\n\n--------FOR RANDOM FOREST CLASSIFIER MODEL-------")
print('Training Accuracy', np.mean(yt_p==y_train))
print('Validation  Accuracy', np.mean(yv_p==y_val))
conf_matrix_train = metrics.confusion_matrix(y_train, yt_p, labels=train_labels)
conf_matrix_val = metrics.confusion_matrix(y_val, yv_p, labels=val_labels)

print('Training confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_train, index=train_labels, columns=train_labels)))
print('Validation confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_val, index=val_labels, columns=val_labels)))



--------FOR KNN MODEL-------
Training Accuracy 0.6528571428571428
Validation  Accuracy 0.5166666666666667
Training confusion matrix:
      0   1    2
0  211  22   23
1   69  69   52
2   49  28  177

Validation confusion matrix:
     0   1   2
0  70  10  23
1  34  18  30
2  21  27  67



--------FOR LOGISTIC REGRESSION MODEL-------
Training Accuracy 0.53
Validation  Accuracy 0.55
Training confusion matrix:
      0  1    2
0  215  0   41
1  119  0   71
2   98  0  156

Validation confusion matrix:
     0  1   2
0  89  0  14
1  49  0  33
2  39  0  76



--------FOR RANDOM FOREST CLASSIFIER MODEL-------
Training Accuracy 1.0
Validation  Accuracy 0.5566666666666666
Training confusion matrix:
      0    1    2
0  256    0    0
1    0  190    0
2    0    0  254

Validation confusion matrix:
     0   1   2
0  68  14  21
1  31  18  33
2  19  15  81



Normalising the dataset and calculating the accuracy again

In [93]:
mean = X_train.mean(0)
sd =  X_train.std(0)

X_train = (X_train-mean)/sd
X_val  = (X_val-mean)/sd

train_labels = np.unique(y_train)
val_labels = np.unique(y_val)


model  = KNeighborsClassifier()
model.fit(X_train,y_train)
yt_p = model.predict(X_train)
yv_p = model.predict(X_val)

print("\n\n--------FOR KNN MODEL-------")
print('Training Accuracy', np.mean(yt_p==y_train))
print('Validation  Accuracy', np.mean(yv_p==y_val))
conf_matrix_train = metrics.confusion_matrix(y_train, yt_p, labels=train_labels)
conf_matrix_val = metrics.confusion_matrix(y_val, yv_p, labels=val_labels)

print('Training confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_train, index=train_labels, columns=train_labels)))
print('Validation confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_val, index=val_labels, columns=val_labels)))


model = LogisticRegression()
model.fit(X_train,y_train)
yt_p = model.predict(X_train)
yv_p = model.predict(X_val)

print("\n\n--------FOR LOGISTIC REGRESSION MODEL-------")
print('Training Accuracy', np.mean(yt_p==y_train))
print('Validation  Accuracy', np.mean(yv_p==y_val))
conf_matrix_train = metrics.confusion_matrix(y_train, yt_p, labels=train_labels)
conf_matrix_val = metrics.confusion_matrix(y_val, yv_p, labels=val_labels)

print('Training confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_train, index=train_labels, columns=train_labels)))
print('Validation confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_val, index=val_labels, columns=val_labels)))

model = RandomForestClassifier()
model.fit(X_train,y_train)
yt_p = model.predict(X_train)
yv_p = model.predict(X_val)

print("\n\n--------FOR RANDOM FOREST CLASSIFIER MODEL-------")
print('Training Accuracy', np.mean(yt_p==y_train))
print('Validation  Accuracy', np.mean(yv_p==y_val))
conf_matrix_train = metrics.confusion_matrix(y_train, yt_p, labels=train_labels)
conf_matrix_val = metrics.confusion_matrix(y_val, yv_p, labels=val_labels)

print('Training confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_train, index=train_labels, columns=train_labels)))
print('Validation confusion matrix:\n {}\n'.format(pd.DataFrame(conf_matrix_val, index=val_labels, columns=val_labels)))



--------FOR KNN MODEL-------
Training Accuracy 0.6257142857142857
Validation  Accuracy 0.54
Training confusion matrix:
      0   1    2
0  216  23   17
1   83  54   53
2   48  38  168

Validation confusion matrix:
     0   1   2
0  75  11  17
1  34  15  33
2  20  23  72



--------FOR LOGISTIC REGRESSION MODEL-------
Training Accuracy 0.5385714285714286
Validation  Accuracy 0.56
Training confusion matrix:
      0  1    2
0  199  8   49
1  102  4   84
2   77  3  174

Validation confusion matrix:
     0  1   2
0  83  0  20
1  38  4  40
2  33  1  81



--------FOR RANDOM FOREST CLASSIFIER MODEL-------
Training Accuracy 1.0
Validation  Accuracy 0.5433333333333333
Training confusion matrix:
      0    1    2
0  256    0    0
1    0  190    0
2    0    0  254

Validation confusion matrix:
     0   1   2
0  65  14  24
1  35  15  32
2  17  15  83



The confusion matrix here represents the number of times that an actual value was predicted correctly/incorrectly. The diagonal of the confusion matrix has the values of True Positives.

For example if we see the confusion matrix of Logistic Regression model, in validation set, out of 300, 168 (83+4+81) times the label was predicted correctly, while it was predicted incorrectly the rest of the times

We can see from the results that Random Forest Classifier is clearly overfitting. Although KNN is giving good training accuracy, the validation accuracy has dropped. Thus it is also slightly overfitting. The logistic regression model is giving a fine accuracy both before and after normalisation and is not overfitting as the training and validation accuracies are quiet near. Thus the best choice here is Logistic Regression Model.

# 9 Conclusions

Here, we will select the Logistic Regression model, because it is giving a fine accuracy and minimum difference between the training and validation accuracy as compared to the other models i.e. overfitting is minimum.

Improvements:

There is a possibility of change in accuracy if we increase the sample size. Thus we can try to train the models by passing larger sample sizes and then compare the  respective accuracy of both models.

We can also change the categorization of spots. This might also increase the accuracy of the model.

Although, we have tried to take the most relevant features of the audio files,  we can try taking other features based on which the accuracy of the model can be improved.